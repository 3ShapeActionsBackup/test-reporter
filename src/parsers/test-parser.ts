import {Endpoints} from '@octokit/types'

type OutputParameters = Endpoints['POST /repos/:owner/:repo/check-runs']['parameters']['output']

export type ParseTestResult = (content: string) => Promise<TestResult>

export interface TestResult {
  success: boolean
  output: OutputParameters
}
