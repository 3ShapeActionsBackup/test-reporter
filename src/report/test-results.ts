export class TestRunResult {
  constructor(readonly suites: TestSuiteResult[], private totalTime?: number) {}

  get tests(): number {
    return this.suites.reduce((sum, g) => sum + g.tests, 0)
  }

  get passed(): number {
    return this.suites.reduce((sum, g) => sum + g.passed, 0)
  }
  get failed(): number {
    return this.suites.reduce((sum, g) => sum + g.failed, 0)
  }
  get skipped(): number {
    return this.suites.reduce((sum, g) => sum + g.skipped, 0)
  }

  get time(): number {
    return this.totalTime ?? this.suites.reduce((sum, g) => sum + g.time, 0)
  }

  get result(): TestExecutionResult {
    return this.suites.some(t => t.result === 'failed') ? 'failed' : 'success'
  }
}

export class TestSuiteResult {
  constructor(readonly name: string, readonly groups: TestGroupResult[], private totalTime?: number) {}

  get tests(): number {
    return this.groups.reduce((sum, g) => sum + g.tests.length, 0)
  }

  get passed(): number {
    return this.groups.reduce((sum, g) => sum + g.passed, 0)
  }
  get failed(): number {
    return this.groups.reduce((sum, g) => sum + g.failed, 0)
  }
  get skipped(): number {
    return this.groups.reduce((sum, g) => sum + g.skipped, 0)
  }
  get time(): number {
    return this.totalTime ?? this.groups.reduce((sum, g) => sum + g.time, 0)
  }

  get result(): TestExecutionResult {
    return this.groups.some(t => t.result === 'failed') ? 'failed' : 'success'
  }
}

export class TestGroupResult {
  constructor(readonly name: string | undefined | null, readonly tests: TestCaseResult[]) {}

  get passed(): number {
    return this.tests.reduce((sum, t) => (t.result === 'success' ? sum + 1 : sum), 0)
  }
  get failed(): number {
    return this.tests.reduce((sum, t) => (t.result === 'failed' ? sum + 1 : sum), 0)
  }
  get skipped(): number {
    return this.tests.reduce((sum, t) => (t.result === 'skipped' ? sum + 1 : sum), 0)
  }
  get time(): number {
    return this.tests.reduce((sum, t) => sum + t.time, 0)
  }

  get result(): TestExecutionResult {
    return this.tests.some(t => t.result === 'failed') ? 'failed' : 'success'
  }
}

export class TestCaseResult {
  constructor(readonly name: string, readonly result: TestExecutionResult, readonly time: number) {}
}

export type TestExecutionResult = 'success' | 'skipped' | 'failed' | undefined
