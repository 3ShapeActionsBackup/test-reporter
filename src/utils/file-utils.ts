import * as fs from 'fs'

export function getFileContent(path: string): string {
  if (!fs.existsSync(path)) {
    throw new Error(`File '${path}' not found`)
  }

  if (!fs.lstatSync(path).isFile()) {
    throw new Error(`'${path}' is not a file`)
  }

  return fs.readFileSync(path, {encoding: 'utf8'})
}

export function normalizeDirPath(path: string, trailingSeparator: boolean): string {
  if (!path) {
    return path
  }

  path = normalizeFilePath(path)
  if (trailingSeparator && !path.endsWith('/')) {
    path += '/'
  }
  return path
}

export function normalizeFilePath(path: string): string {
  if (!path) {
    return path
  }

  return path.trim().replace(/\\/g, '/')
}
