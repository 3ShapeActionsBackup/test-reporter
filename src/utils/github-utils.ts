import * as github from '@actions/github'
import {EventPayloads} from '@octokit/webhooks'

export function getCheckRunSha(): string {
  if (github.context.payload.pull_request) {
    const pr = github.context.payload.pull_request as EventPayloads.WebhookPayloadPullRequestPullRequest
    return pr.head.sha
  }

  return github.context.sha
}
