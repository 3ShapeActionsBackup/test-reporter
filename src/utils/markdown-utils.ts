export enum Align {
  Left = ':---',
  Center = ':---:',
  Right = '---:',
  None = '---'
}

export const Icon = {
  skip: '✖️', // ':heavy_multiplication_x:'
  success: '✔️', // ':heavy_check_mark:'
  fail: '❌' // ':x:'
}

export function details(summary: string, content: string): string {
  return `<details><summary>${summary}</summary>${content}</details>`
}

export function link(title: string, address: string): string {
  return `[${title}](${address})`
}

type ToString = string | number | boolean | Date
export function table(headers: ToString[], align: ToString[], ...rows: ToString[][]): string {
  const headerRow = `| ${headers.map(tableEscape).join(' | ')} |`
  const alignRow = `| ${align.join(' | ')} |`
  const contentRows = rows.map(row => `| ${row.map(tableEscape).join(' | ')} |`).join('\n')
  return [headerRow, alignRow, contentRows].join('\n')
}

export function tableEscape(content: ToString): string {
  return content.toString().replace('|', '\\|')
}
