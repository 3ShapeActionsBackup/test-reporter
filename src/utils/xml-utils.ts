const isoDateRe = /^\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d:[0-5]\d|Z)$/

// matches dotnet duration: 00:00:00.0010000
const durationRe = /^(\d\d):(\d\d):(\d\d\.\d+)$/

export function parseAttribute(str: string | undefined): string | Date | number | undefined {
  if (str === '' || str === undefined) {
    return str
  }

  if (isoDateRe.test(str)) {
    return new Date(str)
  }

  const durationMatch = str.match(durationRe)
  if (durationMatch !== null) {
    const [_, hourStr, minStr, secStr] = durationMatch
    return (parseInt(hourStr) * 3600 + parseInt(minStr) * 60 + parseFloat(secStr)) * 1000
  }

  const num = parseFloat(str)
  if (isNaN(num)) {
    return str
  }

  return num
}
