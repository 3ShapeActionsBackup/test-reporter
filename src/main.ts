import * as core from '@actions/core'
import * as github from '@actions/github'
import {parseJestJunit} from './parsers/jest-junit/jest-junit-parser'
import {ParseTestResult} from './parsers/test-parser'
import {getFileContent} from './utils/file-utils'
import {getCheckRunSha} from './utils/github-utils'

async function run(): Promise<void> {
  try {
    await main()
  } catch (error) {
    core.setFailed(error.message)
  }
}

async function main(): Promise<void> {
  const name = core.getInput('name', {required: true})
  const path = core.getInput('path', {required: true})
  const reporter = core.getInput('reporter', {required: true})
  const token = core.getInput('token', {required: true})

  const octokit = github.getOctokit(token)
  const sha = getCheckRunSha()

  const parser = getParser(reporter)
  const content = getFileContent(path)
  const result = await parser(content)

  await octokit.checks.create({
    head_sha: sha,
    name,
    status: 'completed',
    conclusion: result.success ? 'success' : 'failure',
    output: result.output,
    ...github.context.repo
  })
}

function getParser(reporter: string): ParseTestResult {
  switch (reporter) {
    case 'dotnet-trx':
      throw new Error('Not implemented yet!')
    case 'flutter-machine':
      throw new Error('Not implemented yet!')
    case 'jest-junit':
      return parseJestJunit
    default:
      throw new Error(`Input parameter 'reporter' is set to invalid value '${reporter}'`)
  }
}

run()
