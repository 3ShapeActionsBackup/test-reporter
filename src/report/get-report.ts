import * as core from '@actions/core'
import {TestExecutionResult, TestRunResult, TestSuiteResult} from './test-results'
import {Align, formatTime, Icon, link, table} from '../utils/markdown-utils'
import {slug} from '../utils/slugger'

export interface ReportOptions {
  listSuites?: 'all' | 'failed'
  listTests?: 'all' | 'failed' | 'none'
}

export function getReport(results: TestRunResult[], options: ReportOptions = {}): string {
  core.info('Generating check run summary')

  const maxReportLength = 65535
  const sections: string[] = []

  applySort(results)

  const badge = getReportBadge(results)
  sections.push(badge)

  const runs = getTestRunsReport(results, options)
  sections.push(...runs)

  const report = sections.join('\n')
  if (report.length > maxReportLength) {
    let msg = `**Check Run summary limit of ${maxReportLength} chars was exceed**`
    if (options.listTests !== 'all') {
      msg += '\n- Consider setting `list-tests` option to `only-failed` or `none`'
    }
    if (options.listSuites !== 'all') {
      msg += '\n- Consider setting `list-suites` option to `only-failed`'
    }

    return `${badge}\n${msg}`
  }

  return report
}

function applySort(results: TestRunResult[]): void {
  results.sort((a, b) => a.path.localeCompare(b.path))
  for (const res of results) {
    res.suites.sort((a, b) => a.name.localeCompare(b.name))
  }
}

function getReportBadge(results: TestRunResult[]): string {
  const passed = results.reduce((sum, tr) => sum + tr.passed, 0)
  const skipped = results.reduce((sum, tr) => sum + tr.skipped, 0)
  const failed = results.reduce((sum, tr) => sum + tr.failed, 0)
  return getBadge(passed, failed, skipped)
}

function getBadge(passed: number, failed: number, skipped: number): string {
  const text = []
  if (passed > 0) { text.push(`${passed} passed`) }
  if (failed > 0) { text.push(`${failed} failed`) }
  if (skipped > 0) { text.push(`${skipped} skipped`) }
  let message = text.length > 0 ? text.join(', ') : 'none'

  let color = 'success'
  if (failed > 0) {
    color = 'critical'
  } else if (passed === 0 && failed === 0) {
    color = 'yellow'
  }
  const hint = failed > 0 ? 'Tests failed' : 'Tests passed successfully'
  const uri = encodeURIComponent(`tests-${message}-${color}`)
  return `![${hint}](https://img.shields.io/badge/${uri})`
}

function getTestRunsReport(testRuns: TestRunResult[], options: ReportOptions): string[] {
  const sections: string[] = []

  if (testRuns.length > 1) {
    const tableData = testRuns.map((tr, runIndex) => {
      const time = formatTime(tr.time)
      const name = tr.path
      const addr = makeRunSlug(runIndex).link
      const nameLink = link(name, addr)
      const passed = tr.passed > 0 ? `${tr.passed}${Icon.success}` : ''
      const failed = tr.failed > 0 ? `${tr.failed}${Icon.fail}` : ''
      const skipped = tr.skipped > 0 ? `${tr.skipped}${Icon.skip}` : ''
      return [nameLink, passed, failed, skipped, time]
    })

    const resultsTable = table(
      ['Report', 'Passed', 'Failed', 'Skipped', 'Time'],
      [Align.Left, Align.Right, Align.Right, Align.Right, Align.Right],
      ...tableData
    )
    sections.push(resultsTable)
  }

  const suitesReports = testRuns.map((tr, i) => getSuitesReport(tr, i, options)).flat()
  sections.push(...suitesReports)
  return sections
}

function getSuitesReport(tr: TestRunResult, runIndex: number, options: ReportOptions): string[] {
  const sections: string[] = []

  const time = `${(tr.time / 1000).toFixed(3)}s`

  const slug = makeRunSlug(runIndex)
  const nameLink = `<a id="${slug.id}" href="${slug.link}">${tr.path}</a>`
  const icon = getResultIcon(tr.result)
  sections.push(`## ${nameLink} ${icon}`)

  const headingLine2 = `**${tr.tests}** tests were completed in **${time}** with **${tr.passed}** passed, **${tr.failed}** failed and **${tr.skipped}** skipped.`
  sections.push(headingLine2)

  const suites = options.listSuites === 'failed' ? tr.failedSuites : tr.suites
  if (suites.length > 0) {
    const suitesTable = table(
      ['Test suite', 'Passed', 'Failed', 'Skipped', 'Time'],
      [Align.Left, Align.Right, Align.Right, Align.Right, Align.Right],
      ...suites.map((s, suiteIndex) => {
        const tsTime = formatTime(s.time)
        const tsName = s.name
        const skipLink = options.listTests === 'none' || (options.listTests === 'failed' && s.result !== 'failed')
        const tsAddr = makeSuiteSlug(runIndex, suiteIndex).link
        const tsNameLink = skipLink ? tsName : link(tsName, tsAddr)
        const passed = s.passed > 0 ? `${s.passed}${Icon.success}` : ''
        const failed = s.failed > 0 ? `${s.failed}${Icon.fail}` : ''
        const skipped = s.skipped > 0 ? `${s.skipped}${Icon.skip}` : ''
        return [tsNameLink, passed, failed, skipped, tsTime]
      })
    )
    sections.push(suitesTable)
  }

  if (options.listTests !== 'none') {
    const tests = suites
      .map((ts, suiteIndex) => getTestsReport(ts, runIndex, suiteIndex, options))
      .flat()

    if (tests.length > 1) {
      sections.push(...tests)
    }
  }

  return sections
}

function getTestsReport(ts: TestSuiteResult, runIndex: number, suiteIndex: number, options: ReportOptions): string[] {
  const groups = options.listTests === 'failed' ? ts.failedGroups : ts.groups
  if (groups.length === 0) {
    return []
  }

  const sections: string[] = []

  const tsName = ts.name
  const tsSlug = makeSuiteSlug(runIndex, suiteIndex)
  const tsNameLink = `<a id="${tsSlug.id}" href="${tsSlug.link}">${tsName}</a>`
  const icon = getResultIcon(ts.result)
  sections.push(`### ${tsNameLink} ${icon}`)

  const headingLine2 = `**${ts.tests}** tests were completed in **${ts.time}ms** with **${ts.passed}** passed, **${ts.failed}** failed and **${ts.skipped}** skipped.`
  sections.push(headingLine2)

  for (const grp of groups) {
    const tests = options.listTests === 'failed' ? grp.failedTests : grp.tests
    if (tests.length === 0) {
      continue
    }
    const grpHeader = grp.name ? `\n**${grp.name}**` : ''
    const testsTable = table(
      ['Result', 'Test', 'Time'],
      [Align.Center, Align.Left, Align.Right],
      ...grp.tests.map(tc => {
        const name = tc.name
        const time = formatTime(tc.time)
        const result = getResultIcon(tc.result)
        return [result, name, time]
      })
    )

    sections.push(grpHeader, testsTable)
  }

  return sections
}

function makeRunSlug(runIndex: number): {id: string; link: string} {
  // use prefix to avoid slug conflicts after escaping the paths
  return slug(`r${runIndex}`)
}

function makeSuiteSlug(runIndex: number, suiteIndex: number): {id: string; link: string} {
  // use prefix to avoid slug conflicts after escaping the paths
  return slug(`r${runIndex}s${suiteIndex}`)
}

function getResultIcon(result: TestExecutionResult): string {
  switch (result) {
    case 'success':
      return Icon.success
    case 'skipped':
      return Icon.skip
    case 'failed':
      return Icon.fail
    default:
      return ''
  }
}
