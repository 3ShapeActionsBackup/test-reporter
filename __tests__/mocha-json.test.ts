import * as fs from 'fs'
import * as path from 'path'

import {MochaJsonParser} from '../src/parsers/mocha-json/mocha-json-parser'
import {ParseOptions} from '../src/test-parser'
import {getReport} from '../src/report/get-report'
import {normalizeFilePath} from '../src/utils/path-utils'

describe('mocha-json tests', () => {
  it('produces empty test run result when there are no test cases', async () => {
    const fixturePath = path.join(__dirname, 'fixtures', 'empty', 'mocha-json.json')
    const filePath = normalizeFilePath(path.relative(__dirname, fixturePath))
    const fileContent = fs.readFileSync(fixturePath, {encoding: 'utf8'})

    const opts: ParseOptions = {
      parseErrors: true,
      trackedFiles: []
    }

    const parser = new MochaJsonParser(opts)
    const result = await parser.parse(filePath, fileContent)
    expect(result.tests).toBe(0)
    expect(result.result).toBe('success')
  })

  it('report from ./reports/mocha-json test results matches snapshot', async () => {
    const fixturePath = path.join(__dirname, 'fixtures', 'mocha-json.json')
    const outputPath = path.join(__dirname, '__outputs__', 'mocha-json.md')
    const filePath = normalizeFilePath(path.relative(__dirname, fixturePath))
    const fileContent = fs.readFileSync(fixturePath, {encoding: 'utf8'})

    const opts: ParseOptions = {
      parseErrors: true,
      trackedFiles: ['test/main.test.js', 'test/second.test.js', 'lib/main.js']
    }

    const parser = new MochaJsonParser(opts)
    const result = await parser.parse(filePath, fileContent)
    expect(result).toMatchSnapshot()

    const report = getReport([result])
    fs.mkdirSync(path.dirname(outputPath), {recursive: true})
    fs.writeFileSync(outputPath, report)
  })
})
