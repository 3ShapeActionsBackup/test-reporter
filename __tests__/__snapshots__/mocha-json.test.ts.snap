// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mocha-json tests report from ./reports/mocha-json test results matches snapshot 1`] = `
TestRunResult {
  "path": "fixtures/mocha-json.json",
  "suites": Array [
    TestSuiteResult {
      "groups": Array [
        TestGroupResult {
          "name": "Test 1",
          "tests": Array [
            TestCaseResult {
              "error": undefined,
              "name": "Passing test",
              "result": "success",
              "time": 0,
            },
          ],
        },
        TestGroupResult {
          "name": "Test 1 Test 1.1",
          "tests": Array [
            TestCaseResult {
              "error": Object {
                "details": "Error: Some error
    at Object.throwError (lib\\\\main.js:2:9)
    at Context.<anonymous> (test\\\\main.test.js:15:11)
    at processImmediate (internal/timers.js:461:21)",
                "line": 2,
                "message": "Some error",
                "path": "lib/main.js",
              },
              "name": "Exception in target unit",
              "result": "failed",
              "time": 0,
            },
            TestCaseResult {
              "error": Object {
                "details": "AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:

false !== true

    at Context.<anonymous> (test\\\\main.test.js:11:14)
    at processImmediate (internal/timers.js:461:21)",
                "line": 11,
                "message": "Expected values to be strictly equal:

false !== true
",
                "path": "test/main.test.js",
              },
              "name": "Failing test",
              "result": "failed",
              "time": 1,
            },
          ],
        },
        TestGroupResult {
          "name": "Test 2",
          "tests": Array [
            TestCaseResult {
              "error": Object {
                "details": "Error: Some error
    at Context.<anonymous> (test\\\\main.test.js:22:11)
    at processImmediate (internal/timers.js:461:21)",
                "line": 22,
                "message": "Some error",
                "path": "test/main.test.js",
              },
              "name": "Exception in test",
              "result": "failed",
              "time": 0,
            },
          ],
        },
      ],
      "name": "test/main.test.js",
      "totalTime": undefined,
    },
    TestSuiteResult {
      "groups": Array [
        TestGroupResult {
          "name": null,
          "tests": Array [
            TestCaseResult {
              "error": undefined,
              "name": "Skipped test",
              "result": "skipped",
              "time": 0,
            },
            TestCaseResult {
              "error": Object {
                "details": "Error: Timeout of 1ms exceeded. For async tests and hooks, ensure \\"done()\\" is called; if returning a Promise, ensure it resolves. (C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-reporter\\\\reports\\\\mocha\\\\test\\\\second.test.js)
    at listOnTimeout (internal/timers.js:554:17)
    at processTimers (internal/timers.js:497:7)",
                "line": undefined,
                "message": "Timeout of 1ms exceeded. For async tests and hooks, ensure \\"done()\\" is called; if returning a Promise, ensure it resolves. (C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-reporter\\\\reports\\\\mocha\\\\test\\\\second.test.js)",
                "path": undefined,
              },
              "name": "Timeout test",
              "result": "failed",
              "time": 8,
            },
          ],
        },
      ],
      "name": "test/second.test.js",
      "totalTime": undefined,
    },
  ],
  "totalTime": 12,
}
`;
